### ISSUE TITLE
Bugs in rendering Attributes at Apiary.io

### ISSUE THREAD

**Issue by jmdacruz** [apiaryio/api-blueprint#191]
When defining data structures which contain not only attributes, but also nested data structures, the "Attributes" table at Apiary.io does not show the nested data structures or attributes (it only displays the name of the top-level structure, with type "object")

For example:

- top_level_item1 (string) - A top level item
- top_level_item2 (string) - Another top level item
- properties (object) - A nested data structure
    - item1 (string) - A nested item
    - item2 (string) - Another nested item

Will display a table like this, which doesn't cover all of the MSON data model:

Name | Type | Required | Description | Value
-------- | ------- | ------------ | --------------- | --------
top_level_item1 | string | | A top level item | |
top_level_item2 | string | | A top level item | |
properties | object | | A nested data structure | |

I would expect something like this (or similar):

Name | Type | Required | Description | Value
-------- | ------- | ------------ | --------------- | --------
top_level_item1 | string | | A top level item | |
top_level_item2 | string | | A top level item | |
properties | object | | A nested data structure | |
properties.item1 | string | | A nested item | |
properties.item2 | string | | Another nested item | |

**Comment by tigerpenguin** [comment]
Seeing the same issue, thought I did something wrong. I wish it would show the data structure as well, as nothing under `# Data Structures` is shown on the page. I'd expect it to be under a separate header in the menu, probably called `Data Structures` (or `Objects`), in addition to `Introduction`, `Reference`. 

It would also be beneficial to show the actual type in the attributes table, not object, which would be clickable I guess (which could just take it to the specific data structure page, or expand or something).

Here's an example blueprint.
```
# Group Foo

## FooContainer [/foo_container]

Returns `FooContainer`

+ Attributes (object)
    + description
    + foo (Foo)

# Data Structures

## Foo (object)
+ id
+ bar (Bar)

## Bar (object)
+ id
+ stuff
```

I'd expect the above blueprint to show something like below:

# Group Foo

## FooContainer [/foo_container]

Returns `FooContainer`

### Attributes (object)

Name | Type | Required | Description | Value
---------|--------|--------------|----------------|-------
description ||||
foo | `Foo` |||
foo {id} ||||
foo {bar} | `Bar` |||
foo {bar {id}} ||||
foo {bar {stuff}} ||||

# Data Structures

## Foo (object)
Name | Type | Required | Description | Value
---------|--------|--------------|----------------|-------
id ||||
bar | Bar |||

## Bar (object)
Name | Type | Required | Description | Value
---------|--------|--------------|----------------|-------
id ||||
stuff ||||

**Comment by tigerpenguin** [comment]
Btw, this isn't just an issue with `Data Structures`. `Attributes` table itself only handles first level attributes.

For example,

```
+ Attributes (object)
    + description
    + foo
        + id
        + name
        + price
```
The above will only show the following table:

Name | Type | Required | Description | Value
---------|--------|--------------|-----------------|--------
description ||||
foo ||||

**Comment by zdne** [comment]
Hey @tigerpenguin @jmdacruz ! 
 at the moment the support of MSON in Apiary is in beta stage. We are working on nice, interactive (=foldable) and rich table as well as rendering Data Structure sections itself and the Table of Contents. 

The current simple table is really sparse and basically acts as a placeholder. 

Please bear with us as we working towards bringing the full, rich support for MSON in Apiary. 

Thanks!

Related topics: 

- https://github.com/apiaryio/api-blueprint/issues/25#issuecomment-94598413
- http://stackoverflow.com/questions/29848375/apiblueprint-structures-to-achieve-desired-description

**Comment by jmdacruz** [comment]
Thanks, @zdne!

**Comment by pete001** [comment]
+1

**Comment by foxx** [comment]
Quite excited to see what you guys have planned for this, it's a problem I've been trying to solve for a while now :)

**Comment by rcrooks** [comment]
@zdne I entered a help request around this - please ignore - I hadn't seen this thread. Good to see it's in the works!

**Comment by zdne** [comment]
:+1: I know guys are working on rendering this in Apiary documentation properly, please stay tuned. 

Sneak peek (please note this is not the final version):

![screen shot 2015-05-13 at 11 27 20 am](https://cloud.githubusercontent.com/assets/613617/7618088/96865f8a-f964-11e4-9231-8535da60cb49.png)

**Comment by foxx** [comment]
*drools* that looks really nice

**Comment by jmdacruz** [comment]
That's very nice indeed!

**Comment by jakejgordon** [comment]
+1

**Comment by zdne** [comment]
@Baggz seems there is a lot of people interested in the progress of this in Apiary :ski

**Comment by klevak** [comment]
+1 Major interest at @llbean too. We need to do a better job describing our payloads and jsonSchema isn't exacly "designed for humans".

**Comment by acoulon99** [comment]
+1 From our group as well :)

**Comment by tigerpenguin** [comment]
Thanks @zdne. Can't wait till this gets shipped! Would `parameters` also support interactive rich tables for nested data playload as well? It would be awesome if it can!

**Comment by zdne** [comment]
Thanks @tigerpenguin! There is no plan for changing the parameters rendering at the moment. However in a future we want to converge these two so it will make sense to unify the rendering as well..

**Comment by anta** [comment]
+1 Can't wait to see Attributes section working and MSON support. Really nice feature!!

**Comment by mliberty** [comment]
+1

**Comment by sensever** [comment]
+1

**Comment by JeroenVdb** [comment]
+1

**Comment by underyx** [comment]
+1, would be *so* happy if this went live soon.

Any updates/estimated timeframes for release?

**Comment by zdne** [comment]
@underyx I will check the status out and keep you updated.

**Comment by zdne** [comment]
@underyx 

The Apiary documentation has just been updated to match the visual style of Attributes: 

![pasted image at 2015_06_30 11_26](https://cloud.githubusercontent.com/assets/613617/8428312/18dbcde2-1f1e-11e5-8a69-0331b87cbb58.png)

However this has still the same functionality (e.g. no nested objects). I will share more on this tomorrow or the day after tomorrow.

**Comment by underyx** [comment]
@zdne Neat! Thanks for being so quick to respond!

**Comment by klevak** [comment]
@zdne  noticing that the JSON Schema is also showing up now.  But it's odd how it shows up above the actual response.  In day or two will we see the nested attributes?

**Comment by zdne** [comment]
@klevak Hey Kris I will pass your remark on schema. This was just released so there maybe be still some changes.

>  In day or two will we see the nested attributes?

I wish I could promise this but it got unfortunately bumped by about two weeks due to importance of other things. Sorry about this :disappointed:

**Comment by JeroenVdb** [comment]
Is this issue related (and will it be fixed) to the issue we discovered and described here: https://github.com/apiaryio/drafter/issues/47 ?
Nested Objects aren't shown in the JSON view either.

**Comment by zdne** [comment]
@jakejgordon I believe this might be an unrelated bug – parhaps @pksunkara can clarify?

**Comment by pksunkara** [comment]
Yes, as I have mentioned in apiaryio/drafter#47, it is an unrelated bug.

**Comment by zdne** [comment]
Thanks for clarification @pksunkara !

**Comment by juniorleota** [comment]
Awesome, the description for the first level attributes is viewable today, looks great guys, +1

**Comment by foxx** [comment]
Hey @zdne, any update on when this functionality will be available?

**Comment by zdne** [comment]
@foxx unfortunately I do not have any new ETA to share at the moment. It is one of the top priorities but due to the importance of other things this was postponed :disappointed:

**Comment by jamesramsay** [comment]
This would be brilliant. I was beginning an experiment to move our docs to MSON and this would make my life much much better!

**Comment by klevak** [comment]
@zdne ... all eyes on this one.  I think the anticipation might kill me.  
![capture](https://cloud.githubusercontent.com/assets/914791/8984281/9a1be134-369e-11e5-9498-7ce091dab61e.PNG)

**Comment by zdne** [comment]
I know Kris :/ please stay tuned we are doing what we can!

**Comment by darkaeon10** [comment]
Looking forward to the release of this feature! :D

**Comment by jk2K** [comment]
+1

**Comment by jamesramsay** [comment]
@zdne any update on this?

**Comment by zdne** [comment]
@jamesramsay between a react components and refracts nothing really. All available hands working on it.

**Comment by imZack** [comment]
+1 we are waiting for this :+1:

**Comment by zdne** [comment]
Note I've renamed this issue from "MSON: Nested data structures and attributes are not shown on "Attributes" table" – I hope this will make the scope of this issue more obvious (and thus prevents duplicates).

**Comment by azmorf** [comment]
+1, waiting for this too :+1:

**Comment by fernandosure** [comment]
+1 Please let me know once this is done, so i can update my docs :+1:

**Comment by jcieslar** [comment]
:+1: What's the status of this issue?

**Comment by cdel46** [comment]
+1 Anxiously pacing like an expectant father :+1:

**Comment by ghola** [comment]
+1

**Comment by Randex** [comment]
+1
Why does it take so long?

**Comment by zdne** [comment]
@Randex we are taking this very seriously and working at the maximum to get this delivered. The MSON – as any other data description / schema – is very complex in its essence especially when you consider rendering deeply structures with multiple samples and mutually exclusive properties. 

Please bear with as are we developing this – I believe you won't be disappointed.

**Comment by gigafied** [comment]
:+1:

**Comment by hims10** [comment]
:thumbsup:

**Comment by nickarnold** [comment]
:+1:

**Comment by admorphit** [comment]
:+1:

**Comment by zdne** [comment]
```apib
- message: Hello everyone, I have great news to share with you today! (string)

    Message for all the good folks at the [API Blueprint issue #191](https://github.com/apiaryio/api-blueprint/issues/191)

- author
    - name: Z
    - twitter: @zdne
    - address
        - company: `Apiary, INC`
        - street: 325 Ninth Street (required)
        - city: San Francisco
        - state: CA
        - zip (number, required)
        - tags: red, green, blue - Tags associated with this address
```

![mson](https://cloud.githubusercontent.com/assets/613617/11093107/835e1fb6-8888-11e5-88b2-e2b2f31d8945.png)


We have just released the `attributes-kit` – a library that renders MSON virtually anywhere. This library is the corner stone of MSON rendering in the Apiary. With it, we are getting really close to the proper support of MSON in Apiary. 

I would like to ask you to check out the `attributes-kit` playground at https://attributes-kit.herokuapp.com or use the it on your own and share with us any feedback, bug reports or comments you will find.  

Check out the library at https://github.com/apiaryio/attributes-kit and let us know how you like it!

Big thank you for the waiting and your support. Let’s get the data modeling and visualization right together!

**Comment by pete001** [comment]
Yayyyy :+1:

**Comment by admorphit** [comment]
Cannot seem to get following
```
# Data Structures

## NewPerson
  - Include Person
  - address (string)

## Person (object, fixed)
  - first_name
  - last_name
  - address (object)
```
to display correctly in `attributes-kit` playground.

Thoughts?

**Comment by pksunkara** [comment]
@admorphit Unfortunately complete data structures support is not yet finished.

**Comment by mkuprionis** [comment]
:+1: Great news! Waiting eagerly:)

**Comment by vitorreis** [comment]
:+1:

**Comment by vitorreis** [comment]
+1

**Comment by gigafied** [comment]
@zdne any update on this landing in Apiary?

**Comment by zdne** [comment]
@gigafied Apiary had to unfortunately slow down the works on this due to importance of other tasks. I would personally very much like to see this in January, hope it will work out. Sorry about the delay. 

Meanwhile we have released a new version of parser that also renders JSON Schemas see [here](https://github.com/apiaryio/drafter/releases/tag/v2.1.0) and [here](https://github.com/apiaryio/protagonist/releases/tag/v1.2.1).

**Comment by toobulkeh** [comment]
:+1:

**Comment by o5** [comment]
@zdne are you planning to show `Members` (enumeration of possible values) in `Response / Attributes` table?

I tried something like this:
```
FORMAT: 1A

# Test API

# TEST [GET /client/{uuid}/status]

+ Parameters
    + id (enum[string])

        Id of a Post

        + Members
            + `A` - This is character A
            + `B`
            + `C` - This is character C

+ Response 200 (application/json)
    
    + Attributes
        + id (enum[string])

            Id of a Post
    
            + Members
                + `A` - This is character A
                + `B`
                + `C` - This is character C
        
    + Body
    
            {
                "test": "test"
            }
```

and expected `Possible values:  A , B , C .` also in `Response / Attributes `

![screen shot 2016-01-04 at 18 14 34](https://cloud.githubusercontent.com/assets/218562/12095532/8047c924-b30f-11e5-8bf4-b84f61778486.png)

Maybe I'm doing something wrong :)

Thanks.

**Comment by zdne** [comment]
@o5 the answer is yes. We are also going to unify the rendering of URI parameters and body attributes in near future.

**Comment by Greyvend** [comment]
:+1: Waiting to see this feature on Apiary!

**Comment by CodeHex** [comment]
+1

**Comment by reingroot** [comment]
Looking forward to this feature! +1 :)

**Comment by ericpopivker** [comment]
is there a workaround for this using Blueprint Editor?

This is one feature that is blocking us from using apiary.io at this time.
It is really mandatory if you do anything but simplest flat API.

**Comment by Kratos-SA** [comment]
+1

**Comment by mightbejosh** [comment]
+1

**Comment by adparker** [comment]
@zdne Wondering if you guys can comment on this. Thanks!

**Comment by pksunkara** [comment]
Hey guys, We are working very hard on this currently. Please know that getting this out is our utmost priority. Please stay tuned. Thanks.

**Comment by podolinek** [comment]
+1

**Comment by netmilk** [comment]
Hey guys,

The better rendering of Attributes in Apiary is going to be released soon (days) fixing lot of issues mentioned above. I would love to roll it out to all of you first and I would love if you may provide some feedback if possible. If you're interested please send me your email address used for registration in Apiary via DM on Twitter (@ntmlk) or mail adam@apiary.io.

**Comment by groll75** [comment]
What ever happened to this?. I loved the preview thats was shown in the above thread by that was from last year and haven't seen this roll out yet.. waiting eagerly for this functionality.

**Comment by zdne** [comment]
Hey @groll75 Per Adam's comment above you can ask for access to the early version of the renderer at Apiary. 

The latest development in the rendering component can be followed here: :https://github.com/apiaryio/attributes-kit/releases

**Comment by thomask** [comment]
+1

**Comment by paraskakis** [comment]
Hey everyone. The improved rendering has been released to 100% of the Apiary user base, so no need to ask us to flip the feature switch anymore! Hope you like it and please keep the feedback coming.

**Comment by ConradLang** [comment]
Hi Emmanuel

Has this been turned on for the SEEK team as I don't seem to see the nested attributes (e.g. The thirdParties attribute in our API):
http://docs.adposting.apiary.io/#reference/0/api-operations/create-an-advertisement

Cheers,
Conrad.

**Comment by ConradLang** [comment]
Hi Emmanuel

I've re-saved the document (no changes) and it seems to have updated the public view of the documentation.

Cheers,
Conrad.

From: TOC <clang@seek.com.au<mailto:clang@seek.com.au>>
Date: Thursday, 28 April 2016 at 3:43 PM
To: apiaryio/api-blueprint <reply@reply.github.com<mailto:reply@reply.github.com>>
Subject: Re: [apiaryio/api-blueprint] Bugs in rendering Attributes at Apiary.io (#191)

Hi Emmanuel

Has this been turned on for the SEEK team as I don't seem to see the nested attributes (e.g. The thirdParties attribute in our API):
http://docs.adposting.apiary.io/#reference/0/api-operations/create-an-advertisement

Cheers,
Conrad.

**Comment by paraskakis** [comment]
Conrad,

Please contact us on support [at] apiary.io and we'll help you out.

**Comment by ConradLang** [comment]
Looks like you need to edit your documentation and save before the new rendering is applied.

**Comment by emre-yildirim** [comment]
up 👍

**Comment by thangbxtonish** [comment]
Up (y)
